core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala

```diff
@@ -507,86 +518,126 @@ trait EntityComponent {
+    private def recursiveGetEntityReferenceIds(direction: RecursionDirection, entityIds: Set[Long], accumulatedEntities: Map[Long, AttributeEntityReference], accumulatedPaths: Map[Seq[AttributeEntityReference], Long]): ReadAction[Seq[Seq[AttributeEntityReference]]] = {
+      def oneLevelDown(idBatch: Set[Long]): ReadAction[Set[(Long, EntityRecord)]] = {
+        val query = entityAttributeQuery filter (_.ownerId inSetBind idBatch) join
+          this on { (attr, ent) => attr.valueEntityRef === ent.id && ! attr.deleted } map (rec => (rec._1.ownerId, rec._2))
         query.result.map(_.toSet)
       }
 
-      def oneLevelUp(idBatch: Set[Long]): ReadAction[Set[Long]] = {
-        val query = entityAttributeQuery filter (_.valueEntityRef.isDefined) filter (_.valueEntityRef inSetBind idBatch) join
-          this on { (attr, ent) => attr.ownerId === ent.id && ! ent.deleted } map (_._2.id)
+      def oneLevelUp(idBatch: Set[Long]): ReadAction[Set[(Long, EntityRecord)]] = {
+        val query = entityAttributeQuery filter (_.valueEntityRef inSetBind idBatch) join
+          this on { (attr, ent) => attr.ownerId === ent.id && ! ent.deleted } map (rec => (rec._1.valueEntityRef.get, rec._2))
         query.result.map(_.toSet)
       }
 
       // need to batch because some RDBMSes have a limit on the length of an in clause
       val batchedEntityIds: Iterable[Set[Long]] = createBatches(entityIds)
 
-      val batchActions: Iterable[ReadAction[Set[Long]]] = direction match {
+      val batchActions: Iterable[ReadAction[Set[(Long, EntityRecord)]]] = direction match {
         case Down => batchedEntityIds map oneLevelDown
         case Up => batchedEntityIds map oneLevelUp
       }
 
-      DBIO.sequence(batchActions).map(_.flatten.toSet).flatMap { refIds =>
-        val untraversedIds = refIds -- accumulatedIds
+      DBIO.sequence(batchActions).map(_.flatten.toSet).flatMap { refRecords =>
+        val newAccumulatedEntities = accumulatedEntities ++ refRecords.map(rec => rec._2.id -> rec._2.toReference)
+
+        val newAccumulatedPaths = refRecords.flatMap { rec =>
+          val keysThatEndWithParent = accumulatedPaths.keys.filter(_.last.equals(accumulatedEntities(rec._1)))
+          val pathsToExtend = accumulatedPaths.filter(p => keysThatEndWithParent.toSeq.contains(p._1))
+          val extendedPaths = pathsToExtend.map(p => (p._1 :+ rec._2.toReference, rec._2.id))
+
+          accumulatedPaths ++ extendedPaths
+        }.toMap
+
+        val untraversedIds = refRecords.map(_._2.id) -- accumulatedEntities.keys
         if (untraversedIds.isEmpty) {
-          DBIO.successful(accumulatedIds)
+          DBIO.successful(accumulatedPaths.keys.toSeq)
         } else {
-          recursiveGetEntityReferenceIds(direction, untraversedIds, accumulatedIds ++ refIds)
+          recursiveGetEntityReferenceIds(direction, untraversedIds, newAccumulatedEntities, newAccumulatedPaths)
```

Comment:
I took a stab at rewriting this because the block computing newAccumulatedPaths is not right. Also you should not need the accumulatedEntities parameter.

```
    private def recursiveGetEntityReferenceIds(direction: RecursionDirection, entityIds: Set[Long], accumulatedPaths: Map[Seq[AttributeEntityReference], Long]): ReadAction[Seq[Seq[AttributeEntityReference]]] = {
      def oneLevelDown(idBatch: Set[Long]): ReadAction[Set[(Long, EntityRecord)]] = {
        val query = entityAttributeQuery filter (_.ownerId inSetBind idBatch) join
          this on { (attr, ent) => attr.valueEntityRef === ent.id && ! attr.deleted } map (rec => (rec._1.ownerId, rec._2))
        query.result.map(_.toSet)
      }

      def oneLevelUp(idBatch: Set[Long]): ReadAction[Set[(Long, EntityRecord)]] = {
        val query = entityAttributeQuery filter (_.valueEntityRef inSetBind idBatch) join
          this on { (attr, ent) => attr.ownerId === ent.id && ! ent.deleted } map (rec => (rec._1.valueEntityRef.get, rec._2))
        query.result.map(_.toSet)
      }

      // need to batch because some RDBMSes have a limit on the length of an in clause
      val batchedEntityIds: Iterable[Set[Long]] = createBatches(entityIds)

      val batchActions: Iterable[ReadAction[Set[(Long, EntityRecord)]]] = direction match {
        case Down => batchedEntityIds map oneLevelDown
        case Up => batchedEntityIds map oneLevelUp
      }

      val accumulatedPathsById = accumulatedPaths.map(e => e._2 -> e._1)
      DBIO.sequence(batchActions).map(_.flatten.toSet).flatMap { parentIdsAndChildRecs =>
        val childPaths = parentIdsAndChildRecs.map { case (parentId, childRec) =>
          (accumulatedPathsById(parentId) :+ childRec.toReference) -> childRec.id
        }.toMap

        val newAccumulatedPaths = accumulatedPaths ++ childPaths
        val untraversedIds = parentIdsAndChildRecs.map { case (_, childRec) => childRec.id } -- accumulatedPathsById.keySet

        if (untraversedIds.isEmpty) {
          DBIO.successful(newAccumulatedPaths.keys.toSeq)
        } else {
          recursiveGetEntityReferenceIds(direction, untraversedIds, newAccumulatedPaths)
        }
      }
    }
```

the "get entity subtrees from a list of entities" test passes... but I am not to confident in that test